collect_app/src/main/java/org/odk/collect/android/activities/GeoPointMapActivity.java

```diff
@@ -130,421 +127,316 @@ public void onCreate(Bundle savedInstanceState) {
-            setupMap(googleMap);
-            locationClient.setListener(this);
-            locationClient.start();
-        });
+    public MapFragment createMapFragment() {
+        String mapSdk = getIntent().getStringExtra(GeneralKeys.KEY_MAP_SDK);
+        return (mapSdk == null || mapSdk.equals(PREF_VALUE_GOOGLE_MAPS)) ?
+            new GoogleMapFragment() : new OsmMapFragment();
     }
 
-    @Override
-    protected void onStop() {
-        locationClient.stop();
+    @Override protected void onStart() {
+        super.onStart();
+        if (map != null) {
+            map.setGpsLocationEnabled(true);
+        }
+    }
+
+    @Override protected void onStop() {
+        // To avoid a memory leak, we have to shut down GPS when the activity
+        // quits for good. But if it's only a screen rotation, we don't want to
+        // stop/start GPS and make the user wait to get a GPS lock again.
+        if (!isChangingConfigurations()) {
+            map.setGpsLocationEnabled(false);
+        }
         super.onStop();
     }
 
+    @Override protected void onSaveInstanceState(Bundle state) {
+        super.onSaveInstanceState(state);
+        state.putParcelable(MAP_CENTER_KEY, map.getCenter());
+        state.putDouble(MAP_ZOOM_KEY, map.getZoom());
+        state.putParcelable(POINT_KEY, map.getMarkerPoint(featureId));
+
+        // Flags
+        state.putBoolean(IS_DRAGGED_KEY, isDragged);
+        state.putBoolean(CAPTURE_LOCATION_KEY, captureLocation);
+        state.putBoolean(FOUND_FIRST_LOCATION_KEY, foundFirstLocation);
+        state.putBoolean(SET_CLEAR_KEY, setClear);
+        state.putBoolean(POINT_FROM_INTENT_KEY, pointFromIntent);
+        state.putBoolean(INTENT_READ_ONLY_KEY, intentReadOnly);
+        state.putBoolean(INTENT_DRAGGABLE_KEY, intentDraggable);
+        state.putBoolean(IS_POINT_LOCKED_KEY, isPointLocked);
+
+        // UI state
+        state.putBoolean(PLACE_MARKER_BUTTON_ENABLED_KEY, placeMarkerButton.isEnabled());
+        state.putBoolean(ZOOM_BUTTON_ENABLED_KEY, zoomButton.isEnabled());
+        state.putBoolean(CLEAR_BUTTON_ENABLED_KEY, clearButton.isEnabled());
+        state.putInt(LOCATION_STATUS_VISIBILITY_KEY, locationStatus.getVisibility());
+        state.putInt(LOCATION_INFO_VISIBILITY_KEY, locationInfo.getVisibility());
+    }
+
+    @Override public void destroy() { }
+
     public void returnLocation() {
-        Intent i = new Intent();
-        if (setClear || (readOnly && latLng == null)) {
-            i.putExtra(FormEntryActivity.LOCATION_RESULT, "");
-            setResult(RESULT_OK, i);
+        String result = null;
 
-        } else if (isDragged || readOnly || locationFromIntent) {
+        if (setClear || (intentReadOnly && featureId == -1)) {
+            result = "";
+        } else if (isDragged || intentReadOnly || pointFromIntent) {
             Timber.i("IsDragged !!!");
```

Comment:
This log is a bit obnoxious, and I'm not sure it's even accurate given the condition above. I know you didn't add it but do you think it can be removed?

Reply:
Sure, I'm happy to remove it.

To be honest, I find the logic related to many of these flags to be extremely confusing, and the flag names themselves are quite strange (`setClear`?!??!).  Some of the renaming and the new comments in this PR are intended as a slight improvement, but really I think it needs a bit of an overhaul.  Next after this PR, I'm planning to fix https://github.com/opendatakit/collect/issues/2805, and I expect the fix will make a lot of this mess go away.